File: 파이썬알고리즘/완전탐색/합이같은부분집합.py
def DFS(idx, sum):
    if idx == n:  # 모든 원소를 고려했을 때
        if sum == (total - sum):  # 현재 부분집합의 합과 나머지 부분집합의 합이 같다면
            return True
        else:
            return False
    else:
        # 현재 원소를 포함시키는 경우
        if DFS(idx + 1, sum + a[idx]):
            return True
        # 현재 원소를 포함시키지 않는 경우
        if DFS(idx + 1, sum):
            return True
    return False

if __name__ == "__main__":
    n = int(input())
    a = list(map(int, input().split()))
    total = sum(a)  # 전체 집합의 합
    if DFS(0, 0):  # DFS 시작, 인덱스 0부터 시작, 초기 합은 0
        print("YES")
    else:
        print("NO")

File: 자바알고리즘/DFS-BFS/합이같은부분집합.java
import java.util.*;

class Main {
  static String answer = "NO";
  static int n, total = 0;
  boolean flag = false;

  public void DFS(int L, int sum, int[] arr){
    if(flag) return;
    if(sum > total / 2) return;
    if(L == n){
      if((total - sum) == sum){
        answer = "YES";
        flag = true;
      }
    } else {
      DFS(L + 1, sum + arr[L], arr); // 여기서 sum을 직접 더해서 새로운 값으로 넘김
      DFS(L + 1, sum, arr); // 기존 sum 값 그대로 넘김
    }
  }

  public static void main(String[] args){
    Main T = new Main();
    Scanner in = new Scanner(System.in);
    n = in.nextInt(); // static 변수를 초기화해야 함
    int[] arr = new int[n];

    for(int i = 0; i < n; i++){
      arr[i] = in.nextInt();
      total += arr[i]; // total 값을 계산
    }
    T.DFS(0, 0, arr);
    System.out.println(answer);
  }
}
